### PR TITLE
RF classification: Support transform and evaluate in a single pass

### PR DESCRIPTION
**Pull request by wbo4958**
This PR only supports transformEvalute for multi-classes.

### CODE REVIEW

File: python/tests/test_random_forest.py [hunk 1]
```diff
@@ -354,6 +356,27 @@ def test_random_forest_classifier(
...
+            transfomred_df = spark_rf_model.transform(test_df)
+            prediction_col_name = spark_rf_model.getPredictionCol()
+
+            # MulticlassClassificationEvaluator requires prediction col with double type
+            transformed_df = transfomred_df.withColumn(
+                prediction_col_name, transfomred_df[prediction_col_name].cast("double")
+            )
+            pyspark_f1_score = evaluator.evaluate(transformed_df)
+
+            assert math.fabs(pyspark_f1_score - spark_cuml_f1_score) < 1e-6
```
Comment (lijinf2) [inline python/tests/test_random_forest.py:378]: transformEvaluate seems equivalent to "transform + evaluate (line 369-line 376)". Is it better to move transformEvaluate from model class to perhaps util.py? transformEvaluate seems not in the method list of Pyspark model class, according to https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.ml.classification.RandomForestClassificationModel.html#pyspark.ml.classification.RandomForestClassificationModel

File: python/src/spark_rapids_ml/core.py [hunk 2]
```diff
@@ -654,7 +672,13 @@ def _out_schema(self, input_schema: StructType) -> Union[StructType, str]:
     def _pre_process_data(
         self, dataset: DataFrame
     ) -> Tuple[DataFrame, List[str], bool]:
-        """Pre-handle the dataset before transform."""
+        """Pre-handle the dataset before transform.
+
+        Please note that, this function just transforms the input column if necessary, and
```
Comment (leewyang) [inline python/src/spark_rapids_ml/core.py:677]: Remove comma, e.g. "Please note that this function..."

File: python/tests/test_random_forest.py (hunk 1 above)
Comment (wbo4958) [reply python/tests/test_random_forest.py:380]: Hmm, good point. But transformEvaluate is a customized API to be used internally and it should be different implementation for different subclass.

File: python/src/spark_rapids_ml/classification.py [hunk 3]
```diff
@@ -282,3 +278,121 @@ def evaluate(
...
+        Optional[
+            Callable[
+                [
+                    Union["cudf.DataFrame", np.ndarray],
+                    Union["cudf.DataFrame", np.ndarray],
+                ],
+                pd.DataFrame,
+            ]
+        ],
+    ]:
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/classification.py]: Done. I use variables to represent these functions.

File: python/src/spark_rapids_ml/core.py [hunk 4]
```diff
@@ -105,6 +105,12 @@
 )
 param_alias = ParamAlias("cuml_init", "handle", "num_cols", "part_sizes", "loop")
 
+# Global parameter used by core and subclasses.
+TransformCategory = namedtuple(
+    "TransformCategory", ("transform", "transform_evaluation")
+)
+transform_category = TransformCategory("transform", "transform_evaluation")
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/core.py]: Done

File: python/src/spark_rapids_ml/core.py [hunk 5]
```diff
@@ -719,17 +735,35 @@ def _transform_udf(pdf_iter: Iterator[pd.DataFrame]) -> pd.DataFrame:
...
+        dataset : :py:class:`pyspark.sql.DataFrame`
+            input dataset.
+
+        Returns
+        -------
+        :py:class:`pyspark.sql.DataFrame`
+            transformed dataset
+        """
+        return self._transform_evaluate(
+            dataset, schema=self._out_schema(dataset.schema)
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/core.py:903]: The path for transform will ignore the evaluate function by replacing it with None.

**Comment by wbo4958** [comment]
build

**Review by leewyang — APPROVED** [review]

**Pull request merged**